Add reminder feature

waffledood opened the pull request:
This feature allows the user to set reminders for specific clients. 

Command format: `remind [INDEX] r/ReminderDetails`, where `[INDEX]` is the index of the client the user wishes to set a reminder for and `ReminderDetails` is the specific reminder the user wants to set for a client.

- [X] Implementation 
- [X] Testing 
- [x] Documentation


Closes #64

howardwhw2 commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -88,6 +90,15 @@ public ReadOnlyAddressBook getAddressBook() {
+    public ObservableList<Person> getReminderPersonList() {
+        ObservableList<Person> reminderList = FXCollections.observableArrayList();
+        ReminderPersons reminderPersons = ReminderPersons.getInstance();
+        @SuppressWarnings("unchecked") HashSet<Person> reminderPersonsList = (HashSet<Person>) reminderPersons.clone();
```
Perhaps add a comment on why you need to `suppresswarnings` here

waffledood replied:
thanks for pointing that out Howard! 

as Bryan had suggested an alternative to getting a copy of the `HashSet`, i won't need the `SupressWarnings()` anymore, will remove it in the next commit

howardwhw2 reviewed (approved):
LGTM Just a comment

Ashuh commented on src/main/java/seedu/address/storage/ReminderPersons.java:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * Returns a shallow copy of the HashSet.
+     */
+    public Object clone() {
+        return reminders.clone();
+    }
```
Would it be simpler to use the copy constructor instead of `clone()`?

cindyangXQ reviewed (approved):
LGTM!

waffledood commented:
Hi guys, need another check & review of my pull request, as I've modified the `RemindCommand`, added a `ReminderWindowCommand`, added unit tests & updated some Ui changes to the `ReminderWindow`! 

Hope my PR is fine, do let me know if there are any changes I need to make

howardwhw2 commented on src/main/java/seedu/address/ui/ReminderPersonListPanel.java:
```diff
@@ -0,0 +1,49 @@
+import seedu.address.model.person.Person;
+
+/**
+ * Panel containing the list of persons.
```
Perhaps add if the panel contains only people with reminders?

howardwhw2 commented on src/main/resources/view/ReminderPersonListPanel.fxml:
```diff
@@ -0,0 +1,8 @@
+<?import javafx.scene.control.ListView?>
+<?import javafx.scene.layout.VBox?>
+
+<VBox xmlns="http://javafx.com/javafx/8" xmlns:fx="http://javafx.com/fxml/1">
```
Noticed you are using javafx v8. Maybe it might be better to standardise it to 11 if you are not using any particular feature from 8?

howardwhw2 reviewed (approved):
LGTM. Only some minor comments

cindyangXQ commented on src/main/java/seedu/address/ui/ReminderPersonCard.java:
```diff
@@ -0,0 +1,126 @@
+package seedu.address.ui;
```
This seems the same as `PersonCard`

cindyangXQ commented on src/main/java/seedu/address/ui/ReminderPersonListPanel.java:
```diff
@@ -0,0 +1,49 @@
+package seedu.address.ui;
```
This seems the same as `PersonListPanel`

howardwhw2 commented:
Commenting it here since it missed it earlier. I think your ReminderWindowCommand is in the wrong package? Its in `logic` instead of `logic.commands`

cindyangXQ commented on src/main/java/seedu/address/ui/ReminderWindow.java:
```diff
@@ -3,10 +3,12 @@
 import java.util.logging.Logger;
```
This is very similar to `FavouriteWindow`. Could abstract out a `SingleColumnPersonListWindow`.

waffledood replied:
thanks for pointing that out, abstracted out `FavouriteWindow` & `ReminderWindow` as `SingleColumnPersonListWindow`, removed `ReminderPersonListPanel` & `ReminderPersonCard` to use `PersonListPanel` & `PersonCard` respectively

waffledood commented:
> Commenting it here since it missed it earlier. I think your ReminderWindowCommand is in the wrong package? It is in `logic` instead of `logic.commands`

ah yes it is, thanks for catching that! 😅

cindyangXQ reviewed (commented):
Some repetitive classes could be modified for better abstraction.

hsiaojietng reviewed (approved):
LGTM
